### PR TITLE
fix: State dump creates dumps at the correct point 

### PR DESCRIPTION
**Pull request by nikurt**
The dump needs to be of the state at the beginning of the last block of the previous epoch.

Add latency metrics to functions `fn obtain_state_part()` and `fn apply_state_part()`.

Improves debugging to be more verbose about which state root is used for dumping and application.

Add a command `neard view-state state-parts read-state-header` to read a state header from the DB.

### CODE REVIEW

**Comment by akhi3030** [comment]
Please ping me when this PR is ready to be merged.

File: nearcore/src/state_sync.rs [hunk 1]
```diff
@@ -368,13 +363,17 @@ fn start_dumping(
...
-    let sync_hash_block = chain.get_block(&sync_hash)?;
-    if runtime.cares_about_shard(None, sync_hash_block.header().prev_hash(), shard_id, false) {
-        assert_eq!(num_shards, sync_hash_block.chunks().len() as u64);
-        let state_root = sync_hash_block.chunks()[shard_id as usize].prev_state_root();
-        let state_root_node = runtime.get_state_root_node(shard_id, &sync_hash, &state_root)?;
+    let sync_prev_header = chain.get_block_header(&sync_hash)?;
+    let sync_prev_hash = sync_prev_header.prev_hash();
+    let prev_sync_block = chain.get_block(&sync_prev_hash)?;
+    if runtime.cares_about_shard(None, prev_sync_block.header().prev_hash(), shard_id, false) {
+        assert_eq!(num_shards, prev_sync_block.chunks().len() as u64);
```
Comment (marcelo-gonzalez) [inline nearcore/src/state_sync.rs:370]: for this whole thing, did you consider just using `Chain::get_state_response_part()`? The logic around finding the right state roots and such is already implemented there, so you could just call that function in the main loop of `state_sync_dump()` if `InProgress` just stored the shard ID, part_id and sync_hash. Unless maybe I'm missing something

**Review by marcelo-gonzalez — APPROVED** [review]

File: nearcore/src/runtime/mod.rs [hunk 2]
```diff
@@ -1270,11 +1274,17 @@ impl RuntimeAdapter for NightshadeRuntime {
...
                     Ok(_) => true,
                     // Storage error should not happen
-                    Err(_) => false,
+                    Err(err) => {
+                        tracing::error!(target: "state-parts", ?err, "State part storage error");
+                        false
+                    }
                 }
             }
             // Deserialization error means we've got the data from malicious peer
```
[the remark this answers is not in the document]
Comment (nikurt) [reply nearcore/src/runtime/mod.rs:1283]: Done

File: nearcore/src/state_sync.rs (hunk 1 above)
Comment (nikurt) [reply nearcore/src/state_sync.rs:370]: Amazing suggestion. Done.

**Comment by nikurt** [comment]
@akhi3030 Please take a look

**Review by akhi3030 — APPROVED** [review]

**Pull request merged**